src/meta/load_balance_policy.h

```diff
@@ -0,0 +1,254 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include "meta_data.h"
+
+namespace dsn {
+namespace replication {
+// disk_tag -> targets(primaries/partitions)_on_this_disk
+typedef std::map<std::string, int> disk_load;
+
+enum class balance_type
+{
+    COPY_PRIMARY = 0,
+    COPY_SECONDARY,
+    MOVE_PRIMARY,
+    INVALID,
+};
+ENUM_BEGIN(balance_type, balance_type::INVALID)
+ENUM_REG(balance_type::COPY_PRIMARY)
+ENUM_REG(balance_type::COPY_SECONDARY)
+ENUM_REG(balance_type::MOVE_PRIMARY)
+ENUM_END(balance_type)
+
+bool calc_disk_load(node_mapper &nodes,
+                    const app_mapper &apps,
+                    app_id id,
+                    const rpc_address &node,
+                    bool only_primary,
+                    /*out*/ disk_load &load);
+const std::string &get_disk_tag(const app_mapper &apps, const rpc_address &node, const gpid &pid);
+std::shared_ptr<configuration_balancer_request>
+generate_balancer_request(const app_mapper &apps,
+                          const partition_configuration &pc,
+                          const balance_type &type,
+                          const rpc_address &from,
+                          const rpc_address &to);
+
+struct flow_path;
+class meta_service;
+class load_balance_policy
+{
+public:
+    load_balance_policy(meta_service *svc);
+    virtual ~load_balance_policy() = 0;
+
+    virtual void balance(bool checker, const meta_view *global_view, migration_list *list) = 0;
+    virtual void register_ctrl_commands();
+    virtual void unregister_ctrl_commands();
+
+protected:
+    void init(const meta_view *global_view, migration_list *list);
+    bool execute_balance(
+        const app_mapper &apps,
+        bool balance_checker,
+        bool balance_in_turn,
+        bool only_move_primary,
+        const std::function<bool(const std::shared_ptr<app_state> &, bool)> &balance_operation);
+    bool is_ignored_app(app_id app_id);
+    bool primary_balance(const std::shared_ptr<app_state> &app, bool only_move_primary);
+    bool move_primary(std::unique_ptr<flow_path> path);
```

Comment:
It seems that there are some functions in `load_balance_policy.h` and `load_balance_policy.cpp`, don't be removed in `greedy_load_balancer.cpp`, such as `move_primary`, will they be removed in your further pull request?

Reply:
Yes.